File: src/2025-02-24-deepclaude-rust/cpp/exercises/practice/complex-numbers/complex_numbers.cpp
#include "complex_numbers.h"

namespace complex_numbers {

Complex::Complex(double real, double imag) : real_(real), imag_(imag) {}

Complex Complex::operator+(const Complex& rhs) const {
    return Complex(real_ + rhs.real_, imag_ + rhs.imag_);
}

Complex Complex::operator-(const Complex& rhs) const {
    return Complex(real_ - rhs.real_, imag_ - rhs.imag_);
}

Complex Complex::operator*(const Complex& rhs) const {
    return Complex(real_ * rhs.real_ - imag_ * rhs.imag_,
                  imag_ * rhs.real_ + real_ * rhs.imag_);
}

Complex Complex::operator/(const Complex& rhs) const {
    double denominator = rhs.real_ * rhs.real_ + rhs.imag_ * rhs.imag_;
    return Complex((real_ * rhs.real_ + imag_ * rhs.imag_) / denominator,
                  (imag_ * rhs.real_ - real_ * rhs.imag_) / denominator);
}

Complex Complex::conj() const {
    return Complex(real_, -imag_);
}

double Complex::abs() const {
    return std::sqrt(real_ * real_ + imag_ * imag_);
}

Complex Complex::exp() const {
    double exp_real = std::exp(real_);
    return Complex(exp_real * std::cos(imag_),
                  exp_real * std::sin(imag_));
}

Complex operator+(double lhs, const Complex& rhs) {
    return Complex(lhs) + rhs;
}

Complex operator-(double lhs, const Complex& rhs) {
    return Complex(lhs) - rhs;
}

Complex operator*(double lhs, const Complex& rhs) {
    return Complex(lhs) * rhs;
}

Complex operator/(double lhs, const Complex& rhs) {
    return Complex(lhs) / rhs;
}

}  // namespace complex_numbers

File: src/2025-02-24-deepclaude-rust/javascript/exercises/practice/complex-numbers/complex-numbers.js
//
// This is only a SKELETON file for the 'Complex Numbers' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class ComplexNumber {
  constructor(real, imag) {
    this._real = real;
    this._imag = imag;
  }

  get real() {
    return this._real;
  }

  get imag() {
    return this._imag;
  }

  add(other) {
    return new ComplexNumber(
      this.real + other.real,
      this.imag + other.imag
    );
  }

  sub(other) {
    return new ComplexNumber(
      this.real - other.real,
      this.imag - other.imag
    );
  }

  div(other) {
    const denominator = other.real ** 2 + other.imag ** 2;
    return new ComplexNumber(
      (this.real * other.real + this.imag * other.imag) / denominator,
      (this.imag * other.real - this.real * other.imag) / denominator
    );
  }

  mul(other) {
    return new ComplexNumber(
      this.real * other.real - this.imag * other.imag,
      this.imag * other.real + this.real * other.imag
    );
  }

  get abs() {
    return Math.sqrt(this.real ** 2 + this.imag ** 2);
  }

  conj() {
    return new ComplexNumber(this.real, -this.imag);
  }

  get exp() {
    const expReal = Math.exp(this.real);
    return new ComplexNumber(
      expReal * Math.cos(this.imag),
      expReal * Math.sin(this.imag)
    );
  }
}
